examples/lighting-app/mtk/src/LEDWidget.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <filogic_led.h>
+
+#include "LEDWidget.h"
+
+void LEDWidget::TimerHandler(TimerHandle_t xTimer)
+{
+    LEDWidget * led_widget = (LEDWidget *) pvTimerGetTimerID(xTimer);
+    led_widget->Toggle();
+}
+
+void LEDWidget::StartTimer(uint32_t aTimeoutInMs)
+{
+    if (xTimerIsTimerActive(mTimer))
+        CancelTimer();
+
+    assert(xTimerChangePeriod(mTimer, aTimeoutInMs / portTICK_PERIOD_MS, 100) == pdPASS);
+}
+
+void LEDWidget::CancelTimer(void)
+{
+    assert(xTimerStop(mTimer, 0) == pdPASS);
+}
+
+void LEDWidget::Init(enum led_id led)
+{
+    filogic_led_init();
+
+    mLed = led;
+
+    mTimer = xTimerCreate(Name(),
+                          1,             // == default timer period (mS)
+                          false,         // no timer reload (==one-shot)
+                          (void *) this, // init timer id = app task obj context
+                          TimerHandler); // timer callback handler
+}
+
+const char * LEDWidget::Name(void)
+{
+    const char * name;
+
+    switch (mLed)
+    {
+    case LED_LIGHT:
+        name = "LED_LIGHT";
+        break;
+    case LED_STATUS:
+        name = "LED_STATUS";
+        break;
+    default:
+        name = "LED UNKNOWN";
+        break;
+    }
+
+    return name;
+}
+
+void LEDWidget::Toggle(void)
+{
+    Set(!mState);
+    StartTimer(mState ? mOn : mOff);
+}
+
+void LEDWidget::DoBlink(void)
+{
+    // turn off then turn on
```

Comment:
This restates the code.
Maybe a better comment: why is this ok? I assume code runs quite fast, so would the false/true even be visible? Why are we doing this instead of just Set(true)? Why is a timer needed?

Reply:
LED blink has been achieved by timer callback which is located at [LEDWidget.cpp:25](https://github.com/matter-mtk/connectedhomeip/blob/master/examples/lighting-app/mtk/src/LEDWidget.cpp#L25).

The reason of the timer is needed is LED should be turn on / off periodically to achieve effect of blink.

And `Set(true)` and `Set(false)` is redundant, should be removed.

I fix it in https://github.com/project-chip/connectedhomeip/pull/22404/commits/dfdc8b62b261304fc334967ffefd25a0db3dfb00